src/protocol/sort/apply.rs

```diff
@@ -1,87 +1,141 @@
+use bitvec::bitvec;
 use embed_doc_image::embed_doc_image;
-use permutation::Permutation;
-
-// TODO #OptimizeLater
-// For now, we are using Permutation crate to implement `apply_inv` and `apply` functions.
-// However this uses usize which is either 32-bit or 64-bit depending on the architecture we are using.
-// In our case, if we are sorting less than 2^32 elements (over 4 billion) 32-bits is sufficient.
-// We probably never need a 64-bit number and is not optimal.
-// It would even be cool to use a u16 if you're sorting less than 65,000 items
-// In future, we should plan to change this code to use u32 or u16 based on number of items
 
 #[embed_doc_image("apply", "images/sort/apply.png")]
 #[embed_doc_image("apply_inv", "images/sort/apply_inv.png")]
+
 /// Permutation reorders (1, 2, . . . , m) into (σ(1), σ(2), . . . , σ(m)).
 /// For example, if σ(1) = 2, σ(2) = 3, σ(3) = 1, and σ(4) = 0, an input (A, B, C, D) is reordered into (C, D, B, A) by σ.
 /// ![Apply steps][apply]
-pub fn apply<T: Copy + Default, S>(mut permutation: Permutation, values: &mut S)
-where
-    S: AsMut<[T]>,
-{
-    permutation.apply_slice_in_place(values);
+pub fn apply<T>(permutation: &[u32], values: &mut [T]) {
```

Comment:
Check it out @akoshelev - I was able to remove ALL the bounds on this trait! No need for even Clone or Default!